include/alpaka/block/shared/st/detail/BlockSharedMemStMemberImpl.hpp

```diff
@@ -0,0 +1,168 @@
+#endif
+
+            //! Give the pointer to an exiting variable
+            //!
+            //! @tparam T type of the variable
+            //! @param id unique id of the variable
+            //! @return nullptr if variable with id not exists
+            template<typename T>
+            T* getVarPtr(uint32_t id) const
+            {
+                // Offset in bytes to the next unaligned meta data header behind the variable.
+                uint32_t off = 0;
+
+                // Iterate over allocated data only
+                while(off < m_allocdBytes)
+                {
+                    // Adjust offset to be aligned
+                    uint32_t const alignedMetaDataOffset
+                        = varChunkEnd<MetaData>(off) - static_cast<uint32_t>(sizeof(MetaData));
+                    ALPAKA_ASSERT_OFFLOAD(
+                        (alignedMetaDataOffset + static_cast<uint32_t>(sizeof(MetaData))) <= m_allocdBytes);
+                    MetaData* metaDataPtr = reinterpret_cast<MetaData*>(m_mem + alignedMetaDataOffset);
+                    off = metaDataPtr->offset;
+
+                    if(metaDataPtr->id == id)
+                        break;
+                }
+
+                // Variable not found.
+                if(off >= m_allocdBytes)
+                    return nullptr;
+
+                return reinterpret_cast<T*>(&m_mem[off - sizeof(T)]);
+            }
+
+            //! Get last allocated variable.
+            template<typename T>
+            T* getLatestVarPtr() const
+            {
+                return reinterpret_cast<T*>(&m_mem[m_allocdBytes - sizeof(T)]);
+            }
+
+        private:
+#if BOOST_COMP_GNUC
+#    pragma GCC diagnostic pop
+#endif
+
+            //! Byte offset to the end of the memory chunk
+            //!
+            //! Calculate bytes required to store a type with a aligned starting offset.
+            //! Start offset to the origin of the user data chunk can be calculated with `result - sizeof(T)`.
+            //! The padding is always before the origin of the user data chunk and can be zero byte.
+            //!
+            //! \tparam T type should fit into the chunk
+            //! \param byteOffset Current byte offset.
+            //! \result Byte offset to the end of the data chunk, relative to m_mem..
+            template<typename T>
+            std::uint32_t varChunkEnd(uint32_t byteOffset) const
+            {
+                static_assert(
+                    core::vectorization::defaultAlignment >= alignof(T),
+                    "Unable to get block shared static memory for types with alignment higher than defaultAlignment!");
+                constexpr std::uint32_t align = static_cast<std::uint32_t>(std::max(TMinDataAlignBytes, alignof(T)));
+                return (byteOffset / align + (byteOffset % align > 0u)) * align + static_cast<uint32_t>(sizeof(T));
```

Comment:
This seems needlessly complex, try:
`(byteOffset + align - 1) / align * align`

And since I have seen this code in multiple places by now, consider moving it into a separate function.

Reply:
both are equivalent where the current implementation has the advantage that it is not overflowing when byte offset is near to 4 billion. But for this implementation, the overflow is not a problem at all.

> And since I have seen this code in multiple places by now, consider moving it into a separate function.

This can be done in a separate PR, this should not be mixed with this fix.

Reply:
Created issue: https://github.com/alpaka-group/alpaka/issues/1261